### ISSUE TITLE
[yul] optimizer may increase gas consumption in code containing many nested for loops that have msize, call etc. as loop condition

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#7472]
## Description

oss-fuzz: issue number 17643

```
{
  for {let i_0} lt(i_0,0x60) {i_0 := add(i_0,0x20)}
  {
    for {let i_1} lt(i_1,0x60) {i_1:=add(i_1,0x20)}
    {
      for {for {let i_2} lt(i_2,0x60) {i_2:=add(i_2,0x20)} { for{} msize() {} {}}} msize() {}
      {}
    }
  }
}
```

is optimized to 

```
{
    let i := 0
    for { } lt(i, 0x60) { i := add(i, 0x20) }
    {
        let i_1 := 0
        for { } lt(i_1, 0x60) { i_1 := add(i_1, 0x20) }
        {
            let i_2 := 0
            for { } lt(i_2, 0x60) { i_2 := add(i_2, 0x20) }
            {
                for { } 1 { }
                { if iszero(msize()) { break } }
            }
            for { } 1 { }
            { if iszero(msize()) { break } }
        }
    }
}
```

which increases code size and trips the fuzzer. The additional instructions are zeroing for-loop init variable, and folding the `msize()` builtin when used as a for-loop conditional into the for-loop via `iszero`

## Environment

- Compiler version: latest develop

**Comment by bshastry** [comment]
This behavior is also seen for for-loop conditions involving low-level calls e.g., `call`, `callcode`, and `delegatecall`

```
{
for { let i_0 := 0 } lt(i_0, 0x60) { i_0 := add(i_0, 0x20) } {
for { let i_1 := 0 } lt(i_1, 0x60) { i_1 := add(i_1, 0x20) } {
for { let i_2 := 0 } lt(i_2, 0x60) { i_2 := add(i_2, 0x20) } {
for {}
call(0x10000000000000000000000000000000000000000000000000, 0x100000000000000000000000000000000000000000000000000, 0x1000000000000000000000000000000000000000000000000000, 0x10000000000000000000000000000000000000000000000000000, 0x100000000000000000000000000000000000000000000000000000, 0x1000000000000000000000000000000000000000000000000000000, 0x10000000000000000000000000000000000000000000000000000000){}
{}
}
}
for { let i_3 := 0 } lt(i_3, 0x60) { i_3 := add(i_3, 0x20) } {
}
}
}
```

**Comment by chriseth** [comment]
Maybe we could turn
```
  for { } 1 { } {
    if A { break }
    B
  }
```
into
```
  if iszero(A) { B }
```

**Comment by bshastry** [comment]
That saves us the `break` statement but introduces the `iszero` expression, so doesn't it undo the benefit?

**Comment by bshastry** [comment]
Also, isn't the explicit for-loop-init variable initialization increasing code size as well?

**Comment by chriseth** [comment]
Replacing a for loop and an if by only an if is certainly a benefit, it vastly reduces control flow complexity.

About the explicit init by zero, we should eliminate that, too, yes.

**Comment by bshastry** [comment]
Ah right, I misread your earlier comment wherein we are completely getting rid of for loop construct...

**Comment by bshastry** [comment]
> Maybe we could turn
> 
> ```
>   for { } 1 { } {
>     if A { break }
>     B
>   }
> ```
> 
> into
> 
> ```
>   if iszero(A) { B }
> ```

Do you mean, turning the first into

```
for {} 1 {}
{
  if iszero(A) { B }
}
```

I'm not sure I follow how you managed to remove the for loop while retaining semantics (of iteration).

**Comment by bshastry** [comment]
@chriseth This is blocking fuzzing. Shall we temporarily increase upper bound on maximum evaluation steps for optimized code (perhaps to 2.0) here

https://github.com/ethereum/solidity/blob/5d906cd58d43c1205580ac213fd35387a9f4ca25/test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp#L121

**Comment by chriseth** [comment]
I don't remember what I meant, it is certainly incorrect the way I wrote it.

If explicitly zero-initializing variables is different than not initializing them, we should either remove that (maybe better option) or count it the same way.

But yeah, you can also make the modification for now.

**Comment by bshastry** [comment]
After experimenting a bit with the upper bound, an increase from 1.5 to 4 times the number of maximum steps for unoptimized code seems to work okay. A factor of <= 3 is not sufficient it seems.

For now, I've changed it to a factor of 4 to continue fuzzing.

**Comment by bshastry** [comment]
> If explicitly zero-initializing variables is different than not initializing them, we should either remove that (maybe better option) or count it the same way.

Currently, we are counting each newly introduced block as an additional "step"

https://github.com/ethereum/solidity/blob/d605f8916787cbea15726d63f36a663b24588b83/test/tools/yulInterpreter/Interpreter.cpp#L153-L160

Therefore, the "increase" in code size (if we were to use step count as a proxy for code size) is not because of the explicit initialization to zero (since we don't count statements as steps) but because of the additional scope introduced by the following optimization pattern

```
for {} msize() {} {}
```

to

```
for {} 1 {}
{
  // additional scope introduced by if statement body block
  if iszero(msize()) { break } 
}
```

In conclusion, I think introduction of additional scope as shown by such inputs only increases EVM bytecode size very marginally. I think what is more relevant here is gas consumption since each additional operation enclosed within nested for loops may accumulate gas costs.

**Comment by chriseth** [comment]
This is weird, the block is the only place where we increment the step counter. Maybe it is not the right metric?

**Comment by chriseth** [comment]
I think originally, the step counter was not introduced to measure some kind of effort, but merely to avoid infinite loops.

**Comment by bshastry** [comment]
> I think originally, the step counter was not introduced to measure some kind of effort, but merely to avoid infinite loops.

Ah yes, true. Using it to measure code size was perhaps an afterthought.
